modules/decision/search_pattern.py

```diff
@@ -0,0 +1,242 @@
+        """
+        Returns the square of the distance to it's target location
+        """
+        return (target_posx - current_position.odometry_data.position.east) ** 2 + (
+            target_posy - current_position.odometry_data.position.north
+        ) ** 2
+
+    def __init__(
+        self,
+        camera_fov_forwards: float,
+        camera_fov_sideways: float,
+        search_height: float,
+        search_overlap: float,
+        current_position_x: float,
+        current_position_y: float,
+        distance_squared_threshold: float,
+        small_adjustment: float,
+    ):
+
+        # Set local constants
+        self.__distance_squared_threshold = distance_squared_threshold
+        self.__small_adjustment = small_adjustment
+
+        # The search will be centred around wherever it is initialized
+        self.__search_origin_x = current_position_x
+        self.__search_origin_y = current_position_y
+
+        # Drone's current target
+        self.__target_posx = self.__search_origin_x
+        self.__target_posy = self.__search_origin_y
+
+        # Initialize the drone to the first position in the search pattern
+        self.__current_square = 1
+        self.__current_side_in_square = 0
+        self.__current_pos_on_side = -1
+
+        # Calculate the gap between positions search_gap_width is the left/right distance
+        # search_gap_depth is the forwards/backwards distance
+        self.__search_width = (
+            2 * search_height * tan((camera_fov_sideways * pi / 180) / 2)
+        )
+        self.__search_depth = (
+            2 * search_height * tan((camera_fov_forwards * pi / 180) / 2)
+        )
+        self.__search_gap_width = self.__search_width * (1 - search_overlap)
+        self.__search_gap_depth = self.__search_depth * (1 - search_overlap)
+
+        self.__new_location = True
+
+        # Calculate positions for first square
+        self.__calculate_square_corners()
+        self.__calculate_side_of_square()
+
+    def __calculate_square_corners(self):
+        """
+        Computes the 4 corners of the current square
+        """
+        # Compute the size of square. First square will be a distance of the minimum of width and
+        # depth, each subsequent square will be a width larger
+        square_size = (
+            min(self.__search_gap_depth, self.__search_gap_width)
+            + (self.__current_square - 1) * self.__search_gap_width
+        )
+
+        # If the depth is less than the width, we apply an offset at each corner to ensure the
+        # entire area is scanned
+        adjustment = 0
+        if self.__search_gap_depth < self.__search_width:
+            adjustment = (self.__search_gap_width - self.__search_gap_depth) / 2
+
+        # Calculate the corners based on the offsets and the search origin. Top left corner is moved
+        # right by search_gap_width as the final side of the square will instead cover that part
+        self.__square_corners = [
```

Comment:
a few options here to make it clear which is the top left, top right, bottom left, bottom right:

1. use an object instead that stores the current state (tl, tr, bl, br) with enums
2. use a dict with an enum class for the keys
3. have these as class variables with a state variable that's assigned enum values

by enums I mean have an object with constants for the possible states so instead of assigning state = "top left", you would have state = enum.TOP_LEFT